packages/collections/persistent/map.pony

```diff
@@ -1,557 +1,235 @@
-
-```
-let empty: Map[String,U32] = Maps.empty[String,U32]() // {}
-// Update returns a new map with the provided key set
-// to the provided value. The old map is unchanged.
-let m2 = m1.update("a", 5) // {a: 5}
-let m3 = m2.update("b", 10) // {a: 5, b: 10}
-let m4 = m3.remove("a") // {b: 10}
-
-// You can create a new map from key value pairs.
-let map = Maps.from[String,U32]([("a", 2), ("b", 3)]) // {a: 2, b: 3}
-```
-"""
+use mut = "collections"
 
 primitive Maps
-  fun val empty[K: (Hashable val & Equatable[K] val), V](): Map[K, V] =>
+  fun val empty[K: (mut.Hashable val & Equatable[K] val), V: Any val]
+  (): Map[K, V] =>
     """
     Return an empty map.
     """
-    MapNode[K, V].empty()
+    Map[K, V]._empty()
 
-  fun val from[K: (Hashable val & Equatable[K] val), V]
+  fun val from[K: (mut.Hashable val & Equatable[K] val), V: Any val]
   (pairs: Array[(K, val->V)]): Map[K, V] ? =>
     """
     Return a map containing the provided key-value pairs.
     """
-    var newMap = empty[K, V]()
+    var m = Map[K, V]._empty()
     for pair in pairs.values() do
-      (let k, let v) = pair
-      newMap = newMap.update(k, v)
+      m = m.update(pair._1, pair._2)
     end
-    newMap
+    m
 
-  fun _last_level(): U32 =>
-    """
-    The maximum depth of the tree used for the map (counting from zero)
-    """
-    4
-
-class val LeafNode[K: (Hashable val & Equatable[K] val), V]
-  let _key: K
-  let _value: val->V
-
-  new val create(k: K, v: val->V) =>
-    _key = k
-    _value = v
+class val Map[K: (mut.Hashable val & Equatable[K] val), V: Any val]
+  """
+  A persistent map based on the Compressed Hash Array Mapped Prefix-tree from 'Optimizing Hash-Array Mapped Tries for Fast and Lean Immutable JVM Collections' by Michael J. Steindorfer and Jurgen J. Vinju
+
+  ## Usage
+  ```
+  let empty: Map[String,U32] = Maps.empty[String,U32]() // {}
+  // Update returns a new map with the provided key set
+  // to the provided value. The old map is unchanged.
+  let m2 = m1.update("a", 5) // {a: 5}
+  let m3 = m2.update("b", 10) // {a: 5, b: 10}
+  let m4 = m3.remove("a") // {b: 10}
+  // You can create a new map from key value pairs.
+  let map = Maps.from[String,U32]([("a", 2), ("b", 3)]) // {a: 2, b: 3}
+  ```
+  """
+  let _root: _Node[K, V]
 
-  fun size(): U64 =>
-    """
-    Return the size of the node.
-    """
-    1
+  new val _empty() =>
+    _root = _Node[K, V].empty(0)
 
-  fun _is_leaf(): Bool =>
-    """
-    Predicate testing if the node is a leaf.
-    """
-    true
+  new val _create(r: _Node[K, V]) =>
+    _root = r
 
-  fun apply(k: K): val->V ? =>
+  fun val apply(k: K): val->V ? =>
     """
     Attempt to get the value corresponding to k.
     """
-    if k == _key then _value else error end
+    _root(k.hash().u32(), k)
 
-  fun _get_with_hash(k: K, hash: U32, level: U32): val->V ? =>
+  fun val size(): U64 =>
     """
-    Attempt to get the value corresponding to k, using hash and level.
+    Return the size of the Map.
```

Comment:
what is size of the map in this case? that's going to be rather unclear for someone reading later.